dali/pipeline/pipeline.cc

```diff
@@ -977,28 +996,20 @@ bool Pipeline::IsDeserializable(const std::string &serialized_pipeline) {
 
 void Pipeline::Shutdown() {
   DeviceGuard dg(device_id_);
-  auto& op_nodes = graph_.GetOpNodes();
-  for (auto & node : op_nodes) {
-    if (!is_input_operator(node)) continue;
-    OperatorBase *op_ptr = &(const_cast<dali::OpNode*>(&node)->InstantiateOperator());
-    switch (node.op_type) {
-      case OpType::CPU: {
-        auto *cpu_op_ptr = dynamic_cast<InputOperator<CPUBackend> *>(op_ptr);
-        assert(cpu_op_ptr);
+  for (auto &[name, node] : input_operators_) {
```

Comment:
nitpick
The `input_operators_` is populated after executor's initialization and this is called from the destructor. It seems a bit brittle, for example, if something throws between intialization and populating of this list.

Reply:
> if something throws between intialization and populating of this list.
Well, Executor's `Build` function should be atomic - i.e. if it throws, it shouldn't leave the executor in a semi-initialized state. There's nothing between Build and DiscoverInputOperators. I can add a comment there that those calls should be back-to-back.

Reply:
Comment would be nice!